### DISCUSSION TITLE
Scope per recuperare company_id

### DISCUSSION THREAD

**Discussion by andreinl** [fattureincloud/api#472] [Q&A] [answered]
Salve,

ho visto che questo argomento è già stato trattato diverse volte ma continuo a ricevere errore 401, Reason: Unauthorized.

Uso Python SDK. Per fare un test veloce uso FastAPI. Il codice è molto semplice, ma non funziona:

```

redirect_uri = 'http://localhost:8000/oauth/'
oauth = OAuth2AuthorizationCodeManager(client_id, client_secret, redirect_uri)

app = FastAPI()

@app.get("/")
def read_root():
    return {"Hello": "World"}

@app.get("/auth")
def read_auth():
    # http://localhost:8000/auth

    scopes = [
        Scope.SETTINGS_ALL,
        # Scope.ISSUED_DOCUMENTS_INVOICES_READ,
    ]

    url = oauth.get_authorization_url(scopes, "EXAMPLE_STATE")

    print('>>>>>> ' + url)
    return RedirectResponse(url)

@app.get("/oauth")
def read_oauth(code: Union[str, None] = None, state: Union[str, None] = None):
    if state == "EXAMPLE_STATE":
        token_obj = oauth.fetch_token(code)
        access_token = token_obj.access_token
        refresh_token = token_obj.refresh_token

        configuration = fattureincloud_python_sdk.Configuration(
            host="https://api-v2.fattureincloud.it"
        )
        configuration.scope = 'https://api-v2.fattureincloud.it/scopes/user.read'

        with fattureincloud_python_sdk.ApiClient(configuration) as api_client:
            api_instance = fattureincloud_python_sdk.UserApi(api_client)

            try:
                api_response = api_instance.list_user_companies()
                # In the result you'll find a list of companies, you must use the "id" field of one of those
                print(api_response.data.companies[0].id)
                pprint(api_response)
            except ApiException as e:
                print("Exception when calling UserApi->list_user_companies: %s\n" % e)

    return {
        'Code': code,
        'State': state,
        'access_token': access_token,
        'refresh_token': refresh_token,
        'expires_in': token_obj.expires_in,
    }

````

**Comment by valmoz** [comment]
Buongiorno @andreinl 
posso chiederle dove ha trovato questa stringa?
`https://api-v2.fattureincloud.it/scopes/user.read`

Le consiglio di verificare il suo codice con quanto descritto qui: [SDK OAuth Helper](https://developers.fattureincloud.it/docs/authentication/code-flow/sdk-oauth-helper#one-initialize-the-oauth2authorizationcodemanager-object)

Grazie
Mauro

**Reply by andreinl** [reply]
Ha ragione. Era uno dei tentativi di definire lo Scope. Comunque anche commentandola il risultato è sempre lo stesso: Unauthorized

**Reply by valmoz** [reply]
Gentilmente, mi descriva il flusso che sta applicando nel suo esempio e i risultati ottenuti nei vari step (le chiederei di troncare o mascherare i token passati per evitare problemi di sicurezza)
Grazie

**Reply by andreinl** [reply]
Ho postato il codice completo. Le uniche righe che mancano sono le righe di vari import e due righe dove viene assegnato client_id e client_secret.

Per prima chiamo la funzione read_auth(). Questa funzione a sua volta fa il redirect verso la pagina oauth (funzione read_oauth())

Nella funzione read_auth() ricevo l'url:
'https://api-v2.fattureincloud.it/oauth/authorize?response_type=code&client_id=XXXXXXXXXXX&redirect_uri=http%3A%2F%2Flocalhost%3A8000%2Foauth%2F&scope=settings%3Aa&state=EXAMPLE_STATE'

Nella funzione read_oauth() ricevo access_token e refresh_token. Token expires in 86400

**Reply by valmoz** [reply, marked as answer]
Credo che le manchi [questa parte](https://developers.fattureincloud.it/docs/authentication/code-flow/sdk-oauth-helper#white_check_mark-perform-an-api-request) dopo aver recuperato il token:

```py
configuration = fattureincloud_python_sdk.Configuration()
configuration.access_token = "PREVIOUSLY_RETRIEVED_ACCESS_TOKEN"
```

Mauro

**Reply by andreinl** [reply]
Grazie! Lei ha assolutamente ragione. Quando stavo copiando i vari valori mi sono accorto anch'io.

Grazie di nuovo!

**Reply by valmoz** [reply]
Grato di esserle stato d'aiuto!
